src/v/storage/index_state.h

```diff
@@ -84,6 +212,187 @@ class index_columns {
+    }
+    size_t size() const noexcept {
+        return std::visit([](auto&& c) { return c.size(); }, _val);
+    }
+
+    // These methods are used by serialization
+    chunked_vector<uint32_t> copy_relative_offset_index() const noexcept {
+        return std::visit(
+          [](auto&& c) { return c.copy_relative_offset_index(); }, _val);
+    }
+    chunked_vector<uint32_t> copy_relative_time_index() const noexcept {
+        return std::visit(
+          [](auto&& c) { return c.copy_relative_time_index(); }, _val);
+    }
+    chunked_vector<uint64_t> copy_position_index() const noexcept {
+        return std::visit(
+          [](auto&& c) { return c.copy_position_index(); }, _val);
+    }
+    void assign_relative_offset_index(chunked_vector<uint32_t> vec) noexcept {
+        std::visit(
+          [v = std::move(vec)](auto&& c) mutable {
+              return c.assign_relative_offset_index(std::move(v));
+          },
+          _val);
+    }
+    void assign_relative_time_index(chunked_vector<uint32_t> vec) noexcept {
+        std::visit(
+          [v = std::move(vec)](auto&& c) mutable {
+              return c.assign_relative_time_index(std::move(v));
+          },
+          _val);
+    }
+    void assign_position_index(chunked_vector<uint64_t> vec) noexcept {
+        std::visit(
+          [v = std::move(vec)](auto&& c) mutable {
+              return c.assign_position_index(std::move(v));
+          },
+          _val);
+    }
+
+    void
+    add_entry(uint32_t relative_offset, uint32_t relative_time, uint64_t pos) {
+        std::visit(
+          [relative_offset, relative_time, pos](auto&& c) {
+              return c.add_entry(relative_offset, relative_time, pos);
+          },
+          _val);
+    }
+
+    /// Pop back one element. This is ineffective with columnar format but
+    /// it's not invoked often and when it is invoked it usually invoked not
+    /// that many times.
+    void pop_back(int n = 1) {
+        std::visit([n](auto&& c) { return c.pop_back(n); }, _val);
+    }
+
+    void shrink_to_fit() {
+        std::visit([](auto&& c) { return c.shrink_to_fit(); }, _val);
+    }
+
+    /// Make deep copy
+    index_columns_mux copy() const {
+        return std::visit(
+          [](auto&& c) { return index_columns_mux(c.copy()); }, _val);
+    }
+
+    friend bool operator==(
+      const index_columns_mux& lhs, const index_columns_mux& rhs) noexcept {
+        if (
+          std::holds_alternative<index_columns>(lhs._val)
+          && std::holds_alternative<index_columns>(rhs._val)) {
+            return std::get<index_columns>(lhs._val)
+                   == std::get<index_columns>(rhs._val);
+        }
+        if (
+          std::holds_alternative<compressed_index_columns>(lhs._val)
+          && std::holds_alternative<compressed_index_columns>(rhs._val)) {
+            return std::get<compressed_index_columns>(lhs._val)
+                   == std::get<compressed_index_columns>(rhs._val);
+        }
```

Comment:
what if we just use normal virtual class instead of union type?

Reply:
it will have very busy interface with a lot of methods, I thought that union type will work better in this case

Reply:
> with a lot of methods

wouldn't it be exactly the same number of methods shared between implementations that we have now? with union type there is still a common interface, right?

Reply:
or is this union type an internal thing to the compressed representation? to be clear, i'm talking about the public interface for index_state. is this union type different than that?

Reply:
```
+    static variant_t initialize(bool compressed) {
+        if (compressed) {
+            return compressed_index_columns();
+        }
+        return index_columns();
+    }
```

right. so i guess i'm unclear about

> it will have very busy interface with a lot of methods,

doesn't both columns types have the same interface? or i missing something?

Reply:
I get rid of variant. Now the index_columns thing has an abstract base class. I also refactored it a bit to minimize number of virtual methods (but there are still a lot of them) and made them more high-level.